### PR TITLE
ByteStringBuilder.putInt/putLong: eliminate closure allocation via SWARUtil VarHandle writes

### PR DESCRIPTION
**Pull request by pjfanning**
`ByteStringBuilder.putInt` and `putLong` passed lambdas to `fillArray`, allocating a closure object on every call. This replaces those closures with direct inlined writes using new `SWARUtil.putInt`/`putLong` helpers.

### `SWARUtil`
- Added `putInt(array, index, value, byteOrder)` and `putLong(array, index, value, byteOrder)` — VarHandle `set`-based writes mirroring the existing `getInt`/`getLong` readers
- Added byte-by-byte fallbacks (`putIntBEWithoutMethodHandle`, `putIntLEWithoutMethodHandle`, `putLongBEWithoutMethodHandle`, `putLongLEWithoutMethodHandle`) for environments without VarHandle support

### `ByteStringBuilder`
- `putInt` and `putLong` now call `ensureTempSize` + `SWARUtil.putInt/putLong` directly, updating `_tempLength`/`_length` inline — same structure as `fillArray` but without the closure

```scala
// Before — allocates a lambda on every call
def putInt(x: Int)(implicit byteOrder: ByteOrder): this.type =
  fillArray(4) { (target, offset) => /* byte shifts */ }

// After — no allocation
def putInt(x: Int)(implicit byteOrder: ByteOrder): this.type = {
  ensureTempSize(_tempLength + 4)
  SWARUtil.putInt(_temp, _tempLength, x, byteOrder)
  _tempLength += 4; _length += 4
  this
}
```

`putFloat` and `putDouble` benefit automatically since they delegate to `putInt`/`putLong`.

### `SWARUtilSpec`
- Added tests for all six new methods (`putInt`, `putIntBEWithoutMethodHandle`, `putIntLEWithoutMethodHandle`, `putLong`, `putLongBEWithoutMethodHandle`, `putLongLEWithoutMethodHandle`) covering specific byte layouts at multiple offsets, boundary values (`Int`/`Long` `MaxValue`/`MinValue`), offset correctness (surrounding bytes not clobbered), and round-trip consistency with the corresponding `get*` fallbacks.

### `ByteStringSpec`
- Added dedicated `putInt` and `putLong` test blocks under `"A ByteStringBuilder"` covering: known byte sequences for both byte orders, boundary values, length assertions, multi-put accumulation, and property-based round-trips against `ByteBuffer` for both big-endian and little-endian.

### CODE REVIEW

**Comment by pjfanning** [comment]
I'll look into creating a JMH benchmark

**Comment by He-Pin** [comment]
nice improvement

**Comment by pjfanning** [comment]
```
With Changes
[info] Benchmark                                Mode  Cnt       Score        Error   Units
[info] ByteString_putInt_Benchmark.putIntBE    thrpt    3  239684.591 ± 102494.674  ops/ms
[info] ByteString_putInt_Benchmark.putIntLE    thrpt    3  272971.614 ±  43025.163  ops/ms
[info] ByteString_putInt_Benchmark.putLongBE   thrpt    3  127419.705 ±   6651.774  ops/ms
[info] ByteString_putInt_Benchmark.putLongLE   thrpt    3  137667.126 ±  25170.882  ops/ms
[info] ByteString_putInt_Benchmark.putShortBE  thrpt    3  266982.432 ±  69828.845  ops/ms
[info] ByteString_putInt_Benchmark.putShortLE  thrpt    3  309618.809 ±   5523.773  ops/ms


Without Changes
[info] Benchmark                                Mode  Cnt       Score       Error   Units
[info] ByteString_putInt_Benchmark.putIntBE    thrpt    3  239961.626 ± 24055.818  ops/ms
[info] ByteString_putInt_Benchmark.putIntLE    thrpt    3  238674.817 ± 47275.478  ops/ms
[info] ByteString_putInt_Benchmark.putLongBE   thrpt    3  131730.867 ±  4642.640  ops/ms
[info] ByteString_putInt_Benchmark.putLongLE   thrpt    3  105625.382 ±  5410.259  ops/ms
[info] ByteString_putInt_Benchmark.putShortBE  thrpt    3  271689.337 ±  7647.398  ops/ms
[info] ByteString_putInt_Benchmark.putShortLE  thrpt    3  260050.052 ± 16497.148  ops/ms
```

File: persistence/src/main/scala/org/apache/pekko/persistence/serialization/SnapshotSerializer.scala [hunk 1]
```diff
@@ -81,13 +81,18 @@ class SnapshotSerializer(val system: ExtendedActorSystem) extends BaseSerializer
     Snapshot(snapshotFromBinary(bytes))
 
   private def headerToBinary(snapshot: AnyRef, snapshotSerializer: Serializer): Array[Byte] = {
-    val out = new ByteArrayOutputStream
-    writeInt(out, snapshotSerializer.identifier)
-
     val ms = migrateManifestIfNecessary(Serializers.manifestFor(snapshotSerializer, snapshot))
```
Comment (He-Pin) [inline persistence/src/main/scala/org/apache/pekko/persistence/serialization/SnapshotSerializer.scala]: PR 标题和描述只说消除 `putInt/putLong` 闭包分配，但实际还包含了 `SnapshotSerializer` 的重构（移除 `ByteArrayOutputStream` + `writeInt`）。虽然字节序检查过是一致的（原始 `writeInt` 写 `i >>> 0, i >>> 8, i >>> 16, i >>> 24` = little-endian，新代码用 `LITTLE_ENDIAN`，`headerFromBinary` 也用 `LITTLE_ENDIAN` 读取），但这个改动应单独一个 PR 或至少在 PR 描述中说明。

File: actor/src/main/scala/org/apache/pekko/util/ByteString.scala [hunk 2]
```diff
@@ -2245,60 +2245,32 @@ final class ByteStringBuilder extends Builder[Byte, ByteString] {
...
   def putShort(x: Int)(implicit byteOrder: ByteOrder): this.type = {
-    if (byteOrder == ByteOrder.BIG_ENDIAN) {
-      this += (x >>> 8).toByte
-      this += (x >>> 0).toByte
-    } else if (byteOrder == ByteOrder.LITTLE_ENDIAN) {
-      this += (x >>> 0).toByte
-      this += (x >>> 8).toByte
-    } else throw new IllegalArgumentException("Unknown byte order " + byteOrder)
+    ensureTempSize(_tempLength + 2)
+    SWARUtil.putShort(_temp, _tempLength, x.toShort, byteOrder)
```
Comment (He-Pin) [inline actor/src/main/scala/org/apache/pekko/util/ByteString.scala]: The `putShort` signature is `def putShort(x: Int)`, but the new implementation calls `x.toShort` before passing to SWARUtil. The original implementation used `x >>> 8` and `x >>> 0` on the full `Int` value.

For values outside `Short` range (e.g. `x = 65535` = 0x0000FFFF), `x.toShort` truncates to `-1` before the bit shifts occur in SWARUtil. While the final byte output happens to be the same for this case, the semantics have changed because the parameter type is `Int` but the behavior is now constrained to `Short` range.

The added tests only cover `Short` range values (via `check { (value: Short) => ... }`), so this truncation behavior is not explicitly tested. Either add tests for out-of-range `Int` values to verify the behavior is preserved, or document this as an intentional semantic change.

File: persistence/src/main/scala/org/apache/pekko/persistence/serialization/SnapshotSerializer.scala (hunk 1 above)
Comment (He-Pin) [inline persistence/src/main/scala/org/apache/pekko/persistence/serialization/SnapshotSerializer.scala:84]: The `SnapshotSerializer` changes refactor `headerToBinary` and the serialization method to use direct byte arrays + `SWARUtil.putInt` instead of `ByteArrayOutputStream` + `writeInt`. This is a significant change to the serialization format logic.

However, no tests are added or modified for `SnapshotSerializer` to verify that:
1. The serialized binary format is identical (same byte order, same structure)
2. Round-trip serialization/deserialization works correctly
3. The `migrateManifestIfNecessary` path still produces correct output

Please add tests for `SnapshotSerializer` to cover these cases, or point to existing tests that would catch regressions here.

File: actor/src/main/scala/org/apache/pekko/util/SWARUtil.scala [hunk 3]
```diff
@@ -235,6 +235,84 @@ private[pekko] object SWARUtil {
...
+   * @param array     the byte array to write to
+   * @param index     the index to write at
+   * @param value     the int value to write
+   * @param byteOrder the byte order to use (big-endian or little-endian)
+   */
+  def putInt(array: Array[Byte], index: Int, value: Int, byteOrder: ByteOrder): Unit = {
+    if (byteOrder == ByteOrder.BIG_ENDIAN) {
+      if (intBeArrayViewSupported) {
+        intBeArrayView.set(array, index, value)
+      } else {
```
Comment (He-Pin) [inline actor/src/main/scala/org/apache/pekko/util/SWARUtil.scala:252]: **Correctness concern:** The `putInt`/`putLong`/`putShort` methods silently treat any non-BIG_ENDIAN byte order as LITTLE_ENDIAN. The original `ByteStringBuilder.putInt` threw `IllegalArgumentException` for unknown byte orders:

```scala
} else throw new IllegalArgumentException("Unknown byte order " + byteOrder)
```

While `ByteOrder` only has BE/LE instances in practice, this is still a behavioral regression. Consider adding an explicit validation or at least documenting this change.

File: bench-jmh/src/main/scala/org/apache/pekko/util/ByteString_putInt_Benchmark.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.concurrent.TimeUnit
+
+import org.openjdk.jmh.annotations._
+import org.openjdk.jmh.infra.Blackhole
+
+@State(Scope.Benchmark)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@BenchmarkMode(Array(Mode.Throughput))
+@Fork(2)
+@Warmup(iterations = 5)
```
Comment (He-Pin) [inline bench-jmh/src/main/scala/org/apache/pekko/util/ByteString_putInt_Benchmark.scala:25]: **Performance concern:** The benchmark only measures the new implementation. There is no baseline `@Benchmark` method using the old `fillArray` approach for before/after comparison.

To validate the performance claim (eliminating closure allocation), the benchmark should either:
1. Include a method that replicates the old `fillArray` pattern for direct comparison, or
2. Document that reviewers should run this benchmark against both `main` and the PR branch.

Without a baseline, the benchmark cannot demonstrate the claimed improvement.

File: actor-tests/src/test/scala/org/apache/pekko/util/ByteStringSpec.scala [hunk 5]
```diff
@@ -2634,6 +2634,203 @@ class ByteStringSpec extends AnyWordSpec with Matchers with Checkers {
         }
       }
     }
+
+    "putInt" when {
```
Comment (He-Pin) [inline actor-tests/src/test/scala/org/apache/pekko/util/ByteStringSpec.scala:2638]: **Test coverage gap:** The PR description states that `putFloat` and `putDouble` "benefit automatically since they delegate to `putInt`/`putLong`". However, no tests are added for `putFloat` or `putDouble` to verify this delegation still works correctly after the refactoring.

Please add at least basic round-trip tests for `putFloat` and `putDouble` against `ByteBuffer` (similar to the existing `putInt`/`putLong` property tests) to catch any regression in the delegation path.

File: actor-tests/src/test/scala/org/apache/pekko/util/SWARUtilSpec.scala [hunk 6]
```diff
@@ -57,6 +57,200 @@ class SWARUtilSpec extends AnyWordSpec with Matchers {
       SWARUtil.getShortBEWithoutMethodHandle(testData, 2) should ===(0x0203.toShort)
       SWARUtil.getShortLEWithoutMethodHandle(testData, 2) should ===(0x0302.toShort)
     }
```
Comment (He-Pin) [inline actor-tests/src/test/scala/org/apache/pekko/util/SWARUtilSpec.scala:59]: **Endianness testing gap:** The round-trip and inverse tests are comprehensive, but the direct layout tests are missing explicit coverage for all-zero (`0`) and all-ones (`-1`) byte patterns:

- `putInt(arr, 0, 0, BE)` should produce `[0x00, 0x00, 0x00, 0x00]`
- `putInt(arr, 0, -1, BE)` should produce `[0xFF, 0xFF, 0xFF, 0xFF]`

Similarly for `putLong` (8 bytes of `0x00` or `0xFF`).

These patterns exercise every byte uniformly and are common edge cases in serialization. The inverse tests cover `-1` as a value, but explicit byte-layout assertions would make it clear the byte order is correct for all-bits-set values.

File: actor/src/main/scala/org/apache/pekko/util/SWARUtil.scala (hunk 3 above)
Comment (pjfanning) [reply actor/src/main/scala/org/apache/pekko/util/SWARUtil.scala:252]: there are only 2 ByteOrders in the Java class - https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/nio/ByteOrder.html

File: bench-jmh/src/main/scala/org/apache/pekko/util/ByteString_putInt_Benchmark.scala (hunk 4 above)
Comment (pjfanning) [reply bench-jmh/src/main/scala/org/apache/pekko/util/ByteString_putInt_Benchmark.scala:25]: this is the case with all our benchmarks - it is not hard to run the benchmark with and without the main source changes

File: actor-tests/src/test/scala/org/apache/pekko/util/ByteStringSpec.scala (hunk 5 above)
Comment (pjfanning) [reply actor-tests/src/test/scala/org/apache/pekko/util/ByteStringSpec.scala:2638]: added more tests

File: actor-tests/src/test/scala/org/apache/pekko/util/SWARUtilSpec.scala (hunk 6 above)
Comment (pjfanning) [reply actor-tests/src/test/scala/org/apache/pekko/util/SWARUtilSpec.scala:59]: added more tests

File: persistence/src/main/scala/org/apache/pekko/persistence/serialization/SnapshotSerializer.scala (hunk 1 above)
Comment (pjfanning) [reply persistence/src/main/scala/org/apache/pekko/persistence/serialization/SnapshotSerializer.scala:84]: #2891 has extra tests and they appear in this PR too

**Review by He-Pin — APPROVED** [review]
lgtm
